precompiles/outposts/osmosis/types.go

```diff
@@ -188,28 +204,63 @@ func CreateOnFailedDeliveryField(receiver string) string {
 	return onFailedDelivery
 }
 
-// ValidateInputOutput validate the input and output tokens used in the Osmosis
-// swap.
+// ValidateInputOutput validates the input and output tokens used in the Osmosis swap.
 func ValidateInputOutput(
-	inputDenom, outputDenom, stakingDenom, portID, channelID string,
+	inputDenom, outputDenom, stakingDenom string,
+	evmosChannel IBCChannel,
 ) error {
 	if outputDenom == inputDenom {
 		return fmt.Errorf(ErrInputEqualOutput, inputDenom)
 	}
 
-	osmoIBCDenom := utils.ComputeIBCDenom(portID, channelID, OsmosisDenom)
+	osmoIBCDenom := utils.ComputeIBCDenom(evmosChannel.PortID, evmosChannel.ChannelID, OsmosisDenom)
 
-	// Check that the input token is evmos or osmo.
-	// This constraint will be removed in future
-	validInputs := []string{stakingDenom, osmoIBCDenom}
-	if !slices.Contains(validInputs, inputDenom) {
-		return fmt.Errorf(ErrInputTokenNotSupported, validInputs)
+	// acceptedTokens are the tokens accepted as input or output of the swap.
+	acceptedTokens := []string{stakingDenom, osmoIBCDenom}
+
+	// Check that the input token is aevmos or uosmo.
+	if !slices.Contains(acceptedTokens, inputDenom) {
+		return fmt.Errorf(ErrDenomNotSupported, acceptedTokens)
+	}
+
+	// Check that the output token is aevmos or uosmo.
+	if !slices.Contains(acceptedTokens, outputDenom) {
+		return fmt.Errorf(ErrDenomNotSupported, acceptedTokens)
 	}
 
 	return nil
 }
 
-// SwapPacketData is an utility structure used to wrap args reiceived by the
+// ConvertToOsmosisRepresentation returns the Osmosis representation of the denom from the Evmos
+// representation of aevmos and uosmo. Return an error if the denom is different from one these two.
+func ConvertToOsmosisRepresentation(
+	denom, stakingDenom string,
+	evmosChannel, osmosisChannel IBCChannel,
+) (denomOsmosis string, err error) {
+	osmoIBCDenom := utils.ComputeIBCDenom(
+		evmosChannel.PortID,
+		evmosChannel.ChannelID,
+		OsmosisDenom,
+	)
+
+	switch denom {
+	case osmoIBCDenom:
+		denomOsmosis = OsmosisDenom
+	case stakingDenom:
+		denomPrefix := transfertypes.GetPrefixedDenom(
+			osmosisChannel.PortID,
+			osmosisChannel.ChannelID,
+			denom,
+		)
+		denomTrace := transfertypes.ParseDenomTrace(denomPrefix)
+		denomOsmosis = denomTrace.IBCDenom()
```

Comment:
This doesn't seem right, if the input denom is `aevmos` the expected output denom should be `uosmo` not the IBC denom.